File: demos_python/12_color_tool_refactor/color_tool_app.py
from color import Color


class ColorToolApp:
    def __init__(self) -> None:
        self.colors: list[Color] = []

    def add_color(self) -> None:
        name = input("Please enter the color name > ")
        hex_code = input("Please enter the color hex code > ")

        color = Color.new(name, hex_code)
        self.colors.append(color)

    def show_colors(self) -> None:
        for color in self.colors:
            color.print()

File: demos_rust/12_color_tool_refactor/src/color_tool_app.rs
use crate::color::Color;
use std::io::{stdin, stdout, Write};

pub struct ColorToolApp {
    pub colors: Vec<Color>,
}

impl ColorToolApp {
    pub fn new() -> Self {
        Self { colors: Vec::new() }
    }

    // the method is now a static method
    // it does not have access to the instance of the struct
    fn console_input(prompt: &str) -> String {
        let mut command = String::new();

        print!("{}", prompt);
        stdout().flush().unwrap();
        stdin().read_line(&mut command).unwrap();

        command.trim().to_string()
    }

    pub fn add_color(&mut self) {
        // Self refers to the ColorToolApp struct itself.
        let name = Self::console_input("Enter color name: ");
        let hex_code = Self::console_input("Enter hex code: ");
        let color = Color::new(&name, &hex_code);
        self.colors.push(color);
    }

    pub fn show_colors(&self) {
        // the as_slice method is used to convert the vector to a slice
        // so that we can iterate over the colors
        // without the as_slice method, we would get an error
        // because the vector would be moved
        // and we would not be able to use it again
        for color in self.colors.as_slice() {
            color.print();
        }
    }
}
